File: problems/CSES/Range_Queries/subarray_sum_queries/subarray_sum_queries.py
import sys
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

# Russian
'''
Решение для этой задачи логически правильное, но изза того что Python не очень быстрый,
легче написать решение для этоц задачи на C++.
'''

# English
'''
The solution to this problem is logically correct, but since Python isn't very fast,
it's easier to write a solution to this problem in C++.
'''

def make_node(x):
    v = max(0, x)
    return (x, v, v, v)

def merge(a, b):
    sum_a, pref_a, suff_a, best_a = a
    sum_b, pref_b, suff_b, best_b = b

    total = sum_a + sum_b

    pref = max(pref_a, sum_a + pref_b)
    suff = max(suff_b, sum_b + suff_a)
    best = max(best_a, best_b, suff_a + pref_b)

    return (total, pref, suff, best)

class IterativeSegmentTree:
    def __init__(self, arr, merge, identity):
        self.n = len(arr)
        self.size = 1
        while self.size < self.n:
            self.size *= 2

        self.tree = [identity] * (2 * self.size)

        self.merge = merge
        self.identity = identity

        for i in range(self.n):
            self.tree[self.size + i] = arr[i]

        for i in range(self.size - 1, 0, -1):
            self.tree[i] = self.merge(
                self.tree[i * 2],
                self.tree[i * 2 + 1]
            )

    def set(self, idx, value):
        pos = self.size + idx
        self.tree[pos] = value

        pos //= 2
        while pos >= 1:
            self.tree[pos] = self.merge(
                self.tree[2 * pos],
                self.tree[2 * pos + 1]
            )
            pos //= 2

    def query(self, left, right):
        left += self.size
        right += self.size + 1

        res_left = self.identity
        res_right = self.identity

        while left < right:
            if left % 2 == 1:
                res_left = self.merge(res_left, self.tree[left])
                left += 1
            if right % 2 == 1:
                right -= 1
                res_right = self.merge(self.tree[right], res_right)

            left //= 2
            right //= 2

        return self.merge(res_left, res_right)

    def max_subarray(self):
        return self.tree[1][3]

def main():
    n, m = map(int, input().split())
    arr = [make_node(int(x)) for x in input().split()]
    st = IterativeSegmentTree(arr, merge, (0, 0, 0, 0))

    for _ in range(m):
        k, x = map(int, input().split())
        st.set(k - 1, make_node(x))

        print(st.max_subarray())

main()

File: problems/CSES/Range_Queries/subarray_sum_queries/subarray_sum_queries.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define pb push_back

struct Node {
    ll sum;
    ll pref;
    ll suff;
    ll best;
};

Node make_node(ll x) {
    ll v = max(0LL, x);
    return {x, v, v, v};
}

struct Merge {
    Node operator()(const Node& a, const Node& b) const {
        Node res;

        res.sum = a.sum + b.sum;

        res.pref = max(a.pref, a.sum + b.pref);
        res.suff = max(b.suff, b.sum + a.suff);
        res.best = max({a.best, b.best, a.suff + b.pref});

        return res;
    }
};

template <typename T, typename Merge>
class IterativeSegmentTree {
private:
    int n;
    int size;

    vector<T> tree;

    Merge merge;
    T identity;

public:
    IterativeSegmentTree(
        const vector<T>& arr,
        Merge merge,
        T identity
    )
        : n(arr.size()),
          merge(merge),
          identity(identity)
    {
        size = 1;

        while (size < n) {
            size *= 2;
        }

        tree.assign(2 * size, identity);

        // Leaves
        for (int i = 0; i < n; i++) {
            tree[size + i] = arr[i];
        }

        // Build
        for (int i = size - 1; i >= 1; i--) {
            tree[i] = merge(
                tree[i * 2],
                tree[i * 2 + 1]
            );
        }
    }

    void set(int idx, T value) {
        int pos = size + idx;

        tree[pos] = value;

        pos /= 2;

        while (pos >= 1) {
            tree[pos] = merge(
                tree[pos * 2],
                tree[pos * 2 + 1]
            );

            pos /= 2;
        }
    }

    T query(int left, int right) {
        left += size;

        // [left, right] -> [left, right)
        right += size + 1;

        T res_left = identity;
        T res_right = identity;

        while (left < right) {
            if (left % 2 == 1) {
                res_left = merge(
                    res_left,
                    tree[left]
                );

                left++;
            }

            if (right % 2 == 1) {
                right--;

                res_right = merge(
                    tree[right],
                    res_right
                );
            }

            left /= 2;
            right /= 2;
        }

        return merge(res_left, res_right);
    }

    ll max_subarray() const {
        return tree[1].best;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    cin >> n >> m;

    vector<Node> arr(n);
    for (int i = 0; i < n; i++) {
        ll x;
        cin >> x;
        arr[i] = make_node(x);
    }

    Node identity = {0, 0, 0, 0};
    IterativeSegmentTree<Node, Merge> st(arr, Merge{}, identity);

    for (int i = 0; i < m; i++) {
        int k;
        ll x;
        cin >> k >> x;

        st.set(k - 1, make_node(x));

        cout << st.max_subarray() << '\n';
    }

    return 0;
}
